Add edge picking support to the Cellpicker

Julesdevops opened the pull request:
### Context
Currently, the CellPicker can only pick triangles.

### Results
CellPicker will also try to pick edges.

### Changes
- Added a PolyLine data model class.
- Implemented evaluateLocation in vtkLine
- Changes the picking workflow ; now using `vtkPolyData.getCell` and `vtkPolyData.getCellType`

- [x] Documentation and TypeScript definitions were updated to match those changes

### PR and Code Checklist
- [x] [semantic-release](https://github.com/semantic-release/semantic-release) commit messages
- [x] Run `npm run reformat` to have correctly formatted code

### Testing
- [ ] This change adds or fixes unit tests <!-- Tests should be added for new functionality -->
- [ ] Tested environment:
  - **vtk.js**: v19.4.0
  - **OS**: ArchLinux
  - **Browser**: Firefox 99.0.1

Julesdevops commented:
@finetjul

finetjul commented on Sources/Common/DataModel/PolyLine/index.js:
```diff
@@ -0,0 +1,120 @@
+  model.classHierarchy.push('vtkPolyLine');
+  const superClass = { ...publicAPI };
+
+  model.line = vtkLine.newInstance();
```
no need to store in model, can be a private variable (`const line = vtkLine.newInstance()`).

finetjul commented on Sources/Common/DataModel/PolyLine/index.js:
```diff
@@ -0,0 +1,120 @@
+      const pCoords = [0, 0, 0];
+
+      model.line.getPoints().setPoint(0, ...model.points.getPoint(subId));
+      model.line.getPoints().setPoint(1, ...model.points.getPoint(subId + 1));
```
The following should be faster:
`model.line.getPoints().getData().set(model.points.getData().subarray(3 * subId, 3 * (subId + 2))`

daker commented on Sources/Common/DataModel/PolyLine/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+import vtkCell from '../Cell';
+import { IIntersectWithLine } from '../Line';
+
+interface IPolyLineInitialValues { }
```
can i have an export here and extends from `ICellInitialValues` ?

daker commented on Sources/Common/DataModel/PolyLine/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+	 * @param {Vector3} pcoords The parametric coordinates.
+	 */
+	intersectWithLine(t1: number, t2: number, p1: Vector3, p2: Vector3, tol: number, x: Vector3, pcoords: Vector3): IIntersectWithLine;
+
```
Can you reformat the file to indent it correctly?

finetjul commented on Sources/Common/DataModel/PolyLine/index.js:
```diff
@@ -0,0 +1,120 @@
+    model.line.getPoints().setPoint(subId, model.points.getPoint(subId));
+    model.line
+      .getPoints()
+      .setPoint(subId + 1, model.points.getPoint(subId + 1));
```
same perf improvement as above

daker commented on Sources/Common/DataModel/PolyLine/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+	 */
+	intersectWithLine(t1: number, t2: number, p1: Vector3, p2: Vector3, tol: number, x: Vector3, pcoords: Vector3): IIntersectWithLine;
+
+  evaluateLocation(subId: number, pcoords: Vector3, x: Vector3, weights: Vector2): void
```
Can you add documentation for `evaluateLocation` & `getParametricDistance`

finetjul commented on Sources/Rendering/Core/CellPicker/index.js:
```diff
@@ -324,7 +364,12 @@ function vtkCellPicker(publicAPI, model) {
       const point = [];
-      minCell.evaluateLocation(minPCoords, point, weights);
+
+      if (vtkCellTypes.hasSubCells(cellType)) {
```
Instead of cellType, shouldn't it be `minCellType` ? or `minCell.getType()` ? or `getCellType(minCell)`

finetjul commented on Sources/Rendering/Core/CellPicker/index.js:
```diff
@@ -264,30 +271,56 @@ function vtkCellPicker(publicAPI, model) {
+      const triangle = vtkTriangle.newInstance();
+      const line = vtkLine.newInstance();
+      const polyLine = vtkPolyLine.newInstance();
+      const numberOfCells = data.getNumberOfCells();
```
how about creating an object ?
```
const tempCell  = {
  CellType.VTK_LINE: vtkLine.newInstance(),
  CellType.VTK_POLY_LINE: ... };
...
  const cell = tempCell[ cellType];
```

finetjul commented on Sources/Rendering/Core/CellPicker/index.js:
```diff
@@ -239,7 +241,12 @@ function vtkCellPicker(publicAPI, model) {
     let pDistMin = Number.MAX_VALUE;
     const minPCoords = [0, 0, 0];
     let minCellId = -1;
```
let's use `let minCellId // = undefined;` to be consistent with `subId`

Julesdevops commented:
@daker @finetjul Thanks for the feedback, I addressed your comments.
I got rid of `getParametricDistance` override in `vtkPolyLine`, because it seems that it is not needed at all.

finetjul commented on Sources/Rendering/Core/CellPicker/index.js:
```diff
@@ -1,15 +1,25 @@
+    [CellType.VTK_LINE]: vtkLine.newInstance(),
+    [CellType.VTK_POLY_LINE]: vtkPolyLine.newInstance(),
+    [CellType.VTK_TRIANGLE]: vtkTriangle.newInstance(),
+  };
```
`return new Map(CellType.keys().map((cellType) => return [cellType, CellType.getClassNameFromTypeId(cellType)])`

Julesdevops replied:
imo this is elegant, but we can't use it.
We don't support all cells, and we would need to dynamically import the relevant module from the class string, and I don't know how to do this properly without creating problems.
If someday we supports all the cells listed in `vtk-js/Sources/Common/DataModel/CellTypes/Constants.js` and also have a map directly with the corresponding modules we shall use it.

finetjul replied:
fair enough

Julesdevops commented on Sources/Common/DataModel/PolyLine/index.js:
```diff
@@ -0,0 +1,107 @@
+    pcoords,
+    dist2,
+    weights
+  ) => {}; // virtual
```
I don't know. I've done like here: https://github.com/Kitware/vtk-js/blob/master/Sources/Common/DataModel/Line/index.js#L220
If it make sense to add an error message, then I'm adding it to both locations

finetjul replied:
Actually, both Line and PolyLine should not even overwrite that method. Its implementation already exist in vtkCell (which should have the vtkErrorMacro)

Julesdevops replied:
Make sense

finetjul commented on Sources/Common/DataModel/PolyLine/index.js:
```diff
@@ -0,0 +1,98 @@
+  const superClass = { ...publicAPI };
+
+  const line = vtkLine.newInstance();
+  line.getPoints().setNumberOfPoints(2);
```
shouldn't this be done automatically in `vtkLine::extend()`  ?

finetjul commented on Sources/Rendering/Core/CellPicker/index.js:
```diff
@@ -264,30 +276,46 @@ function vtkCellPicker(publicAPI, model) {
     if (locator) {
       // TODO when cell locator will be implemented
     } else if (data.getPolys) {
```
this test is no longer needed

Julesdevops replied:
Which test ? `data.getPolys` or `locator`?

finetjul replied:
data.getPolys

finetjul commented on Sources/Common/DataModel/PolyLine/index.js:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    weights[0] = 1.0 - pcoords[0];
+    weights[1] = pcoords[0];
```
why not do:
```
line
        .getPoints()
        .getData()
        .set(model.points.getData().subarray(3 * subId, 3 * (subId + 2)));

      return line.evaluateLocation(...);
```

Julesdevops replied:
yes, make sense. Thanks

finetjul commented on Sources/Rendering/Core/CellPicker/index.js:
```diff
@@ -263,31 +275,47 @@ function vtkCellPicker(publicAPI, model) {
+        minCellType = data.getCellType(cellId);
+        const cell = tempCellMap[minCellType];
+
+        if (cell === undefined) {
```
`if (cell == null)` is more flexible

Julesdevops replied:
Sure but it will never be null, or am I missing something?
If it's null it's because one of the `newInstance`calls returned null so in this case I think it's better to have an error than silently fail.

finetjul replied:
To me it is not this function responsibility to trigger such error.
However, this function needs to support such scenario and not crash.

Julesdevops commented:
@finetjul all done

daker reviewed (approved):
The TS definitions looks good
